glib/src/main_context_futures.rs

```diff
@@ -149,26 +155,30 @@ impl WakerSource {
     }
 }
 
-unsafe impl Send for TaskSource {}
-unsafe impl Sync for TaskSource {}
+unsafe impl<T> Send for TaskSource<T> {}
+unsafe impl<T> Sync for TaskSource<T> {}
 
 unsafe impl Send for WakerSource {}
 unsafe impl Sync for WakerSource {}
 
-impl TaskSource {
+impl<T: 'static> TaskSource<T> {
+    const TASK_SOURCE_FUNCS: ffi::GSourceFuncs = ffi::GSourceFuncs {
```

Comment:
This was `static` for a reason :) `g_source_new()` just takes a pointer and assumes that this stays valid forever.

Reply:
It seems to be working

Reply:
I think that's just coincidence. If you check the gobject code, `g_source_new()` assumes that the pointer to the functions stays valid forever.

Reply:
I made an example where it shows the compiler does do static monomorphizations of the vtable: https://rust.godbolt.org/z/6fxMT87zs

But this behavior does not seem to be specified anywhere.

Reply:
Yes, `const` are only guaranteed to be like a `#define` in C. The behaviour should be equivalent to copying the value into every place where it is used.

Reply:
Changed it to use `Box<dyn Any>`, that should still avoid allocating if return type is `()`

Reply:
Why the box instead of the plain value? This also seems to be for a different discussion thread but not sure which? :)

Reply:
The `T` parameter cannot be used in a `static` variable so I had to remove it.

Reply:
Ah I see. The `T` would have to propagate up to `TaskSource<T>` and all the `GSourceFuncs`.

Seems like a good solution this way